Move the remainder of control theming code back to RenderTheme from Theme

weinig opened the pull request:
#### a02bdae433235dd1f932e85ebc1bf4fda3e2525b
<pre>
Move the remainder of control theming code back to RenderTheme from Theme
<a href="https://bugs.webkit.org/show_bug.cgi?id=294127">https://bugs.webkit.org/show_bug.cgi?id=294127</a>

Reviewed by Darin Adler.

A long ago effort to create a platform level theme library has
not borne fruit. This moves the remaining control theming code
in platform&apos;s Theme class hierarchy back to WebCore&apos;s RenderTheme
class class hierarchy.

Having it all in one place is good and this will make efforts
to add typed wrappers for layout values more tractable.

* Source/WebCore/platform/Theme.cpp:
* Source/WebCore/platform/Theme.h:
* Source/WebCore/platform/adwaita/ThemeAdwaita.cpp:
* Source/WebCore/platform/adwaita/ThemeAdwaita.h:
* Source/WebCore/platform/mac/ThemeMac.h:
* Source/WebCore/platform/mac/ThemeMac.mm:
* Source/WebCore/rendering/RenderTheme.cpp:
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.cpp:
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.h:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
* Source/WebCore/rendering/mac/RenderThemeMac.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:

Canonical link: <a href="https://commits.webkit.org/295974@main">https://commits.webkit.org/295974@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/62e42c6f5deaa734bcbff30e59900ecc89186d6d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106821 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26572 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16970 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112032 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/57419 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108860 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27242 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35074 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/81114 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109825 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21627 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96394 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61455 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21067 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14502 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56873 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90968 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14530 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114987 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33959 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25052 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/90176 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34324 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92625 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89886 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22931 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34826 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12640 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29644 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33884 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/39297 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33631 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36984 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35229 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -42169,8 +42169,6 @@
 				F3ABFE0C130E9DA000E7F7D1 /* InstrumentingAgents.h in Headers */,
 				2D550A1429C30F3F00BD0C1F /* IntDegrees.h in Headers */,
-				BA79DEF12DDDF89D000DB04C /* IntegrityPolicy.h in Headers */,
-				BA5AA81B2DDF59210038722C /* IntegrityPolicyViolationReportBody.h in Headers */,
```
I have bet it's correct, but what is this change about?

weinig replied:
Xcode did it! I bet it will be gone when I rebase.

darinadler commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1366,32 +1408,33 @@ void RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioStyle
-    if (minControlSize.height.value() > style.minHeight().value())
-        style.setMinHeight(WTFMove(minControlSize.height));
+    auto minimumControlSize = this->minimumControlSize(appearance, style.fontCascade(), { style.minWidth(), style.minHeight() }, { style.width(), style.height() }, style.usedZoom());
+    // FIXME: This check makes no sense for cases when the LengthType of the values are different.
```
Seems like a FIXME++ to me!

weinig replied:
I fix it in the strong style size types change that is also up!! 😇

darinadler commented on Source/WebCore/rendering/mac/RenderThemeMac.mm:
```diff
@@ -825,6 +859,222 @@ static NSControlSize controlSizeFromPixelSize(std::span<const IntSize, 4> sizes,
+
+static const std::array<IntSize, 4>& switchSizes()
+{
+    static const std::array<IntSize, 4> sizes =
```
Maybe we should make `constexpr` work with `IntSize`.

weinig replied:
Good idea!

weinig replied:
It already does work with constexpr. Just need to change these sites.

darinadler commented on Source/WebCore/rendering/mac/RenderThemeMac.mm:
```diff
@@ -825,6 +859,222 @@ static NSControlSize controlSizeFromPixelSize(std::span<const IntSize, 4> sizes,
+
+static void inflateControlPaintRect(StyleAppearance appearance, FloatRect& zoomedRect, float zoomFactor, bool isVertical)
+{
+    BEGIN_BLOCK_OBJC_EXCEPTIONS
```
Where’s the ObjC?

weinig replied:
There is a little below various calls to `supportsLargeFormControls()`, but I kinda doubt that would ever throw. But given this is just a move, I am hesitant to remove it. Someone with AppKit source could probably double check it.

darinadler reviewed (approved):
Looks fine. Wondering why this doesn’t delete `Theme.h` completely yet.

weinig committed:
Move the remainder of control theming code back to RenderTheme from Theme
https://bugs.webkit.org/show_bug.cgi?id=294127

Reviewed by Darin Adler.

A long ago effort to create a platform level theme library has
not borne fruit. This moves the remaining control theming code
in platform's Theme class hierarchy back to WebCore's RenderTheme
class class hierarchy.

Having it all in one place is good and this will make efforts
to add typed wrappers for layout values more tractable.

* Source/WebCore/platform/Theme.cpp:
* Source/WebCore/platform/Theme.h:
* Source/WebCore/platform/adwaita/ThemeAdwaita.cpp:
* Source/WebCore/platform/adwaita/ThemeAdwaita.h:
* Source/WebCore/platform/mac/ThemeMac.h:
* Source/WebCore/platform/mac/ThemeMac.mm:
* Source/WebCore/rendering/RenderTheme.cpp:
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.cpp:
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.h:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
* Source/WebCore/rendering/mac/RenderThemeMac.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:

Canonical link: https://commits.webkit.org/295974@main
